libnmstate/nm/connection.py

```diff
@@ -182,226 +178,247 @@ def get_setting_duplicate(self, setting_name):
+                    error_msg = (
+                        f"Connection {self.profile.get_id()} failed: "
+                        f"state={self._nm_ac.get_state()} "
+                        f"reason={self._nm_ac.get_state_reason()} dev=None"
                     )
+                logging.error(error_msg)
+                self._ctx.fail(
+                    NmstateLibnmError(f"{action} failed: {error_msg}")
                 )
 
-    @staticmethod
-    def _is_connection_unavailable(err):
-        return (
-            isinstance(err, GLib.GError)
-            and err.domain == "nm-manager-error-quark"
-            and err.code == 2
-            and "is not available on the device" in err.message
-        )
-
-    def _get_activation_metadata(self):
-        if self._nmdevice:
-            activation_type = "device"
-            activation_object = self._nmdevice.get_iface()
-        elif self._con_id:
-            activation_type = "connection_id"
-            activation_object = self._con_id
-        else:
-            activation_type = activation_object = "unknown"
-
-        return activation_type, activation_object
-
-    def waitfor_active_connection_async(self, ac):
-        ac.handlers.add(
-            ac.nm_active_connection.connect(
-                "state-changed", self._waitfor_active_connection_callback, ac
+    def _wait_ac_activation(self, action):
+        self._ac_handlers.add(
+            self._nm_ac.connect(
+                "state-changed", self._ac_state_change_callback, action
             )
         )
-        ac.handlers.add(
-            ac.nm_active_connection.connect(
+        self._ac_handlers.add(
+            self._nm_ac.connect(
                 "notify::state-flags",
-                self._waitfor_state_flags_change_callback,
-                ac,
+                self._ac_state_flags_change_callback,
+                action,
             )
         )
-        ac.device_handlers.add(
-            ac.nmdevice.connect(
-                "state-changed", self._waitfor_device_state_change_callback, ac
+
+    def _wait_dev_activation(self, action):
+        if self._nm_dev:
+            self._dev_handlers.add(
+                self._nm_dev.connect(
+                    "state-changed", self._dev_state_change_callback, action
+                )
             )
-        )
 
-    def _waitfor_device_state_change_callback(
-        self, _dev, _new_state, _old_state, _reason, ac
+    def _dev_state_change_callback(
+        self, _dev, _new_state, _old_state, _reason, action,
     ):
-        self._waitfor_active_connection_callback(None, None, None, ac)
+        self._ac_state_change_callback(None, None, None, action)
 
-    def _waitfor_state_flags_change_callback(self, _nm_act_con, _state, ac):
-        self._waitfor_active_connection_callback(None, None, None, ac)
+    def _ac_state_flags_change_callback(self, _nm_act_con, _state, action):
+        self._ac_state_change_callback(None, None, None, action)
```

Comment:
I think instead of `None` all the we should just remove then and do `self._ac_state_change_callback(action)`.

Reply:
All changed to call `self._activation_progress_check(action)`